nomt/src/merkle/page_set.rs

```diff
@@ -15,23 +16,32 @@ pub enum PageOrigin {
     /// It could have been fetched from the hash table, thereby having an associated `BucketInfo`.
     Persisted(BucketInfo),
     /// It could have been reconstructed on the fly without being stored anywhere.
-    /// Keeps track of the total number of leaves in child pages.
-    Reconstructed(u64),
+    /// It keeps track of the total number of leaves in child pages and which nodes
+    /// in the page have been reconstructed.
+    Reconstructed(u64, PageDiff),
 }
 
 impl PageOrigin {
     /// Extract `BucketInfo` from `PageOrigin::Persisted` variant.
     pub fn bucket_info(self) -> Option<BucketInfo> {
         match self {
             PageOrigin::Persisted(bucket_info) => Some(bucket_info),
-            PageOrigin::Reconstructed(_) => None,
+            PageOrigin::Reconstructed(_, _) => None,
         }
     }
 
     /// Extract the number of leaves from `PageOrigin::Reconstructed` variant.
     pub fn leaves_counter(&self) -> Option<u64> {
         match self {
-            PageOrigin::Reconstructed(counter) => Some(*counter),
+            PageOrigin::Reconstructed(counter, _) => Some(*counter),
+            PageOrigin::Persisted(_) => None,
+        }
+    }
+
+    /// Extract the [`PageDiff`] from `PageOrigin::Reconstructed` variant.
```

Comment:
We might as well make it typechecked. Same as above.

```suggestion
    /// Extract the [`PageDiff`] from [`PageOrigin::Reconstructed`] variant.
```